src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.hpp

```diff
@@ -40,12 +68,76 @@ class ShenandoahAdaptiveHeuristics : public ShenandoahHeuristics {
                                                      size_t actual_free);
 
   void record_cycle_start();
+  void record_success_concurrent();
+  void record_success_degenerated();
+  void record_success_full();
 
-  virtual bool should_start_gc() const;
+  virtual bool should_start_gc();
 
   virtual const char* name()     { return "Adaptive"; }
   virtual bool is_diagnostic()   { return false; }
   virtual bool is_experimental() { return false; }
+
+ private:
+  // These are used to adjust the margin of error and the spike threshold
+  // in response to GC cycle outcomes. These values are shared, but the
+  // margin of error and spike threshold trend in opposite directions.
+  const static double FULL_PENALTY_SD;
+  const static double DEGENERATE_PENALTY_SD;
+
+  const static double MINIMUM_CONFIDENCE;
+  const static double MAXIMUM_CONFIDENCE;
+
+  const static double LOWEST_EXPECTED_AVAILABLE_AT_END;
+  const static double HIGHEST_EXPECTED_AVAILABLE_AT_END;
+
+  friend class ShenandoahAllocationRate;
```

Comment:
Does it still need to be friends? I.e. are there non-public members in `ShenandoahAllocationRate` this class needs to access?